src/Database/Esqueleto/Record.hs

```diff
@@ -0,0 +1,423 @@
+    ]
+
+-- | Information about a record we need to generate the declarations.
+-- We compute this once and then pass it around to save on complexity /
+-- repeated work.
+data RecordInfo = RecordInfo
+  { -- | The original record's name.
+    name :: Name
+  , -- | The generated @Sql@-prefixed record's name.
+    sqlName :: Name
+  , -- | The original record's constraints. If this isn't empty it'll probably
+    -- cause problems, but it's easy to pass around so might as well.
+    constraints :: Cxt
+  , -- | The original record's type-variable-binders.
+    typeVarBinders :: [TyVarBndr ()]
+  , -- | The original record's kind, I think.
+    kind :: Maybe Kind
+  , -- | The original record's constructor name.
+    constructorName :: Name
+  , -- | The original record's field names and types, derived from the
+    -- constructors.
+    fields :: [(Name, Type)]
+  , -- | The generated @Sql@-prefixed record's field names and types, computed
+    -- with 'sqlFieldType'.
+    sqlFields :: [(Name, Type)]
+  }
+
+-- | Get a `RecordInfo` instance for the given record name.
+getRecordInfo :: Name -> Q RecordInfo
+getRecordInfo name = do
+  TyConI dec <- reify name
+  let (constraints, typeVarBinders, kind, constructors) =
+        case dec of
+          DataD constraints' _name typeVarBinders' kind' constructors' _derivingClauses ->
+            (constraints', typeVarBinders', kind', constructors')
+          NewtypeD constraints' _name typeVarBinders' kind' constructor' _derivingClauses ->
+            (constraints', typeVarBinders', kind', [constructor'])
+          _ -> error $ "Esqueleto records can only be derived for records and newtypes, but " ++ show name ++ " is neither"
+      constructor = fromJust $ listToMaybe constructors
+      constructorName =
+        case fromJust $ listToMaybe constructors of
+          RecC name' _fields -> name'
+          con -> error $ nonRecordConstructorMessage con
+      fields = getFields constructor
+      sqlFields = toSqlField `map` fields
+      sqlName = makeSqlName name
+
+  pure RecordInfo {..}
+  where
+    getFields :: Con -> [(Name, Type)]
+    getFields (RecC _name fields) = [(fieldName', fieldType') | (fieldName', _bang, fieldType') <- fields]
+    getFields con = error $ nonRecordConstructorMessage con
+
+    toSqlField (fieldName', ty) = (fieldName', sqlFieldType ty)
+
+-- | Create a new name by prefixing @Sql@ to a given name.
+makeSqlName :: Name -> Name
+makeSqlName name = mkName $ "Sql" ++ nameBase name
+
+-- | Transforms a record field type into a corresponding `SqlExpr` type.
+--
+-- * @'Entity' x@ is transformed into @'SqlExpr' ('Entity' x)@.
+-- * @'Maybe' ('Entity' x)@ is transformed into @'SqlExpr' ('Maybe' ('Entity' x))@.
+-- * @x@ is transformed into @'SqlExpr' ('Value' x)@.
+sqlFieldType :: Type -> Type
+sqlFieldType fieldType =
+  case fieldType of
+    -- Entity x -> SqlExpr (Entity x)
+    AppT (ConT ((==) ''Entity -> True)) _innerType -> AppT (ConT ''SqlExpr) fieldType
```

Comment:
i really cannot get over how much i like this use of view patterns to match on a relatively complicated type!